fix(update): add json-merge-patch guidance

toumorokoshi opened the pull request:
After offline discussion, it was agreed upon that IETF RFC 7396 (JSON Merge Patch) is the more sensible choice for a partial update on an HTTP+JSON API. The rationale includes:

- can generally be expressed via protobuf field masks.
- aligns with the AEPs usage of resource-oriented guidance.
- aligns with an existing IETF standard, which as already been widely adopted.

Other minor changes:

- removes duplicative guidance around LROs to make the AEP succinct.
- populates oas guidance, sans etags which may require some more design (in HTTP etags are not added to the resource body).

Fixes #109.

<!--
Describe the big picture of your changes here to communicate to the maintainers
why we should accept this pull request. If it fixes a bug or resolves a feature
request, be sure to link to that issue.
-->

## 🍱 Types of changes

What types of changes does your code introduce to AEP? _Put an `x` in the boxes
that apply_

- [x] Enhancement
- [ ] [New proposal](https://aep.dev/1#workflow)
- [ ] Migrated from google.aip.dev
- [ ] Chore / Quick Fix

## 📋 Your checklist for this pull request

Please review the [AEP Style and Guidance](https://aep.dev/style-guide) for
contributing to this repository.

### General

- [x] Basic [Guidance](https://aep.dev/style-guide#guidance) is met.
- [x] Ensure that your PR
      [references AEPs](https://aep.dev/style-guide#referencing-aeps)
      correctly.
- [x] [My code has been formatted](https://aep.dev/contributing#formatting)
      (usually `prettier -w .`)

### Additional checklist for a new AEP


💝 Thank you!

toumorokoshi commented:
FYI @dv-stewarts if you'd like to take a look and give your thoughts.

gibson042 commented on aep/general/0134/aep.md.j2:
```diff
@@ -27,8 +27,10 @@ Update methods are specified using the following pattern:
+- The method **must** support partial resource update, and the HTTP verb
+  **must** be `PATCH`.
+- If the resources take long than is reasonable for a regular API request, the
+  API **should** use a [long-runing-operation](/long-running-operations).
```
```suggestion
  API **should** use a [long-running-operation](/long-running-operations).
```

gibson042 commented on aep/general/0134/aep.md.j2:
````diff
@@ -57,7 +59,28 @@ separator, `snake_case` is used.
+                $ref: '#/components/schemas/Book'
+```
+
+- The method **must** adhere to the behavior specified in [IETF 7396 - JSON Merge Patch](https://datatracker.ietf.org/doc/html/rfc7396).
````
```suggestion
- The method **must** adhere to the behavior specified in [RFC 7396 - JSON Merge Patch](https://datatracker.ietf.org/doc/html/rfc7396).
```

toumorokoshi replied:
How about IETF RFC? I think that's important to make it clear what body one can find the guidance in.

gibson042 commented on aep/general/0134/aep.md.j2:
````diff
@@ -57,7 +59,28 @@ separator, `snake_case` is used.
+                $ref: '#/components/schemas/Book'
+```
+
+- The method **must** adhere to the behavior specified in [IETF 7396 - JSON Merge Patch](https://datatracker.ietf.org/doc/html/rfc7396).
````
That behavior corresponds with request media type `application/merge-patch+json`, which is not used above. Assuming the generalization to cover `application/json` as well is intentional, I have concerns about forward compatibility with a format that better handles merging arrays (i.e., by actually doing so rather than by replacement) or comprehends `null` differently. And at any rate, the divergence from RFC 7396 merits a specific mention.

toumorokoshi replied:
ah good point - Do you think we'd want to support both `merge-patch+json` and `json`? I'd imagine so from a user perspective - I'm not sure many will recall that detail about merge-patch.

re forward compatibility - are you thinking that if we want to support a new format, we could do so by adding another MIME type? Is there such a format we could start to look at now?

gibson042 replied:
> ah good point - Do you think we'd want to support both `merge-patch+json` and `json`? I'd imagine so from a user perspective - I'm not sure many will recall that detail about merge-patch.

I see no reason *not* to support `application/merge-patch+json` for requesting its behavior; it's extending that behavior to `application/json` that concerns me.

> re forward compatibility - are you thinking that if we want to support a new format, we could do so by adding another MIME type?

Yes, but if `application/json` today means JSON Merge Patch, then everybody is stuck with it forever, despite its obvious flaws.

> Is there such a format we could start to look at now?

Well, [JSON Patch](https://www.rfc-editor.org/rfc/rfc6902) already exists, but is not a great fit for AEP. And I have a draft of a "JSON List Patch" format that fixes JSON Patch's issues with future compatibility in mind, but it's not even public and I don't know if this group wants to pioneer in that direction.

The fundamental issue is that there is more than one JSON-expressible approach to patching a resource and the current state of the art isn't great, but picking a winner for silently specializing `application/json` is a **permanent** decision.

toumorokoshi replied:
Sorry it took me a while to round back.

> The fundamental issue is that there is more than one JSON-expressible approach to patching a resource and the current state of the art isn't great, but picking a winner for silently specializing application/json is a permanent decision.

Fair enough, I do see the point that setting a de-factor winner might not be productive.

It seems like there's universal agreement among those here that the guidance specifying support of json-merge-patch is valuable, so I'm going to shrink this PR to focus on that. We can discuss further formats of support in future PRs. How does that sound?

toumorokoshi replied:
I've updated the guidance and removed mention of `application/json` for now.

dv-stewarts commented on aep/general/0134/aep.md.j2:
```diff
@@ -57,7 +59,28 @@ separator, `snake_case` is used.
+  /v1/publishers/{publisher}/books:
+    patch:
+      summary: Update a book
+      operationId: publishers.book:create
```
There is no guidance for OAS operationId, only general "method name" guidance (which should probably be moved into the proto tab, but that is a different discussion). However, in [AEP-131](https://aep.dev/131) there is such guidance and extrapolating from it, I would think this should be 

```suggestion
      operationId: updateBook
```
It definitely shouldn't be anything with create in it.

toumorokoshi replied:
:thinking: well, I don't know if we've actually made a decision about the operationId yet, and I think that it's usage will be very critical when we start building clients.

For now I'll remove it, and will probably do a follow-up to determine the standard around OAS operations IDs.

toumorokoshi replied:
here's a tracking issue: https://github.com/aep-dev/aep.dev/issues/208

toumorokoshi replied:
apologies for the delay getting to these comments - agreed there is no explicit guidance here, and I've been removing the setting of `operationId` from the examples since there isn't a standard that AEP suggests yet.

If you have a proposal, feel free to add it to issue above!

dv-stewarts commented on aep/general/0134/aep.md.j2:
```diff
@@ -211,6 +203,26 @@ message UpdateBookRequest {
+  /v1/publishers/{publisher}/books:
+    patch:
+      summary: Update a book
+      operationId: publishers.book:create
```
I think this is also wrong.

dv-stewarts commented:
AEP includes [guidance on HTTP and gRPC transcoding](https://aep.dev/127). Will supporting update_mask for gRPC and json-merge-patch for OAS cause the transcoded gRPC APIs to be invalid wrt the OAS spec? Is that a common occurrence and is it viewed as a problem?

We have been using gRPC transcoding as a way to get people into gRPC on the server side without having to fight so many battles with API consumers. Having "native" OAS APIs working differently than transcoded APIs while both claiming to be AEP compliant is likely to sow confusion.

toumorokoshi commented:
@dv-stewarts Thanks for taking a look!

> AEP includes [guidance on HTTP and gRPC transcoding](https://aep.dev/127). Will supporting update_mask for gRPC and json-merge-patch for OAS cause the transcoded gRPC APIs to be invalid wrt the OAS spec? Is that a common occurrence and is it viewed as a problem? 

> We have been using gRPC transcoding as a way to get people into gRPC on the server side without having to fight so many battles with API consumers. Having "native" OAS APIs working differently than transcoded APIs while both claiming to be AEP compliant is likely to sow confusion.


It isn't a very common occurrence - and ideally we wouldn't deviate. There are a lot of tensions w.r.t. this problem overall, including:

- different features in protocols (e.g. protobuf 2 will default non-nullable fields, while in json all fields are nullable).
- HTTP+JSON already has a rich ecosystem of IETF RFCs, many with significant adoption and also not looking like AEP guidance with existing protobufs.

I don't think there's a lot of conflicts generally between the protobuf and HTTP+JSON APIs in the context of the AEPs. and when there are, we try to reconcile them. I don't think we've agreed 100% on how this should be addressed. The conclusion we have reached so far is:

- json-merge-patch is probably the best solution so far (sans yet another standard) for PATCH in JSON.
- field masks are not a great solution for protobuf, but is already widely adopted. 

Ideally, I think we'd replace existing protobuf guidance with something that aligns better with json-merge-patch. However, json-merge-patch uses the resource as the schema, while in protobuf that's not really an option given that the message schema may include non-nullable fields, which mean that you can't really send the intent that "you don't care about the field" without some sort of sidecar (in this case, the field mask).

At Google IIUC json-merge-patch is supported (e.g. https://cloud.google.com/workflows/docs/reference/googleapis/compute/v1/backendBuckets/patch), and there is a transcoding layer that handles the conversion of the patch to the protobuf field mask. 

If you have a specific suggestion, it's definitely up for discussion! I'd encourage you to share any ideas you may have.

toumorokoshi commented:
@mkistler adding you as an official reviewer, thank you!

mkistler commented on aep/general/0134/aep.md.j2:
````diff
@@ -57,7 +59,31 @@ separator, `snake_case` is used.
+```
+
+- The method **must** adhere to the behavior specified in [IETF RFC 7396 - JSON Merge Patch](https://datatracker.ietf.org/doc/html/rfc7396).
+- The method **must** support MIME types `application/merge-patch+json` to adhere to IETF RFC 7396, as well as `application/json` as a common generalization.
````
I wonder if it is a good idea to allow "application/json". RFC7396 has certain requirements, e.g. how null values are processed, that could be disregarded if the body is sent as "application/json". That creates ambiguity which seems bad to me.

toumorokoshi replied:
my primary concern with not supporting `application/json` is that it's very common for users to expect try that first - I suppose they'll just get a 415 unsupported media type or something to that effect and could try again though.

dv-stewarts replied:
Must it support application/json? If this was a "may" then we'd be leaving it to the API implementer to decide if they want to allow this looser type + the default that application/json is a merge-patch.

toumorokoshi replied:
Apologies for the delay here. based on https://github.com/aep-dev/aeps/pull/206#discussion_r1720988023 and your feedback, I'll be removing the optional clause on `application/json` for now - Richard's feedback in the linked thread that we shouldn't make merge-patch the defacto standard here makes sense to me, especially I think we don't feel there's a perfect protocol today. 

I'll resolve this thread, but feel free to chime in further in the linked one!

mkistler commented on aep/general/0134/aep.md.j2:
```diff
@@ -211,6 +206,25 @@ message UpdateBookRequest {
+      requestBody:
+        required: true
+        content:
+          application/json:
```
I think we want "application/merge-patch+json" here, either instead of or in addition to "application/json".

mkistler reviewed (commented):
Left one suggestion for a minor wording change.

I'm also unclear if we should allow "application/json" content type.

Otherwise this looks good.

fsaintjacques commented:
I find it confusing that this would be a JSON first solution where-as this is a protobuf driven spec. I'd much prefer to stick with FieldMask than having to deal with this required infrastructure (transcoding from json-patch to grpc-fieldmask). That almost makes me ignore aep as a potential user.

toumorokoshi commented:
> I find it confusing that this would be a JSON first solution where-as this is a protobuf driven spec. I'd much prefer to stick with FieldMask than having to deal with this required infrastructure (transcoding from json-patch to grpc-fieldmask). That almost makes me ignore aep as a potential user.

Thanks for the feedback! It's worth clarifying - aep is not a proto-driven spec. It is intended to be protocol agnostic principles, with a precise specification in multiple protocols (grpc and HTTP+JSON are the two in scope today). We're working on better docs, but there's a video where the project goals are explained: https://www.youtube.com/watch?v=we9Qp2kHsvY&t=1s.

That all said I do acknowledge the difficulty in building a proper transcoding layer when the interface is so different. I think we'd like to introduce tooling (possibly building adapter services similar to grpc-gateway, or extend / contribute to it). But in the weekly meeting discussions we've agreed that we should be provide idiomatic guidance for each protcol we support, and diverge only when there's complete incompatibility between the two. In this case, we discussed that it's better to choose the widely supported json-merge-patch specification. (But of course, a separate maintainer has to approve this guidance to ratify it).

If there is an alternate idiomatic interface that works well for both, definitely interested to hear of one!

toumorokoshi commented:
I've also added an explanation of json-merge-patch: As an FYI, grpc-HTTP bindings such grpc-gateway support this patch-style translation, so I think ultimately this guidance is actual interoperable with existing proto and proto-http-binding best practices.

toumorokoshi commented:
@gibson042 @dv-stewarts @mkistler sorry for the long delay on this one - I've rebased the content, removed the guidance to suggest `application/json`, and added a blurb explaining why this actually works out ok for both HTTP and proto. PTAL, and I appreciate the dialog!
